docs/adr/0012_construct_as_transformation.md

```diff
@@ -0,0 +1,441 @@
+  - In the `resource.creator` transformation, the `input` field is mapped
+    to a corresponding access type (e.g. `file`, `ociArtifact`).
+- **Access types** (as used in component constructors) depend on whether
+  they are marked as `by value` or `by reference`.
+  - **by value** access types are mapped to a combination of
+    `resource.creator`, `resource.downloader` and `local.resource.uploader`
+    (emphasis on **local**) transformations.
+  - **by reference** access types are mapped to a single
+    `resource.creator` transformation.
+
+**Conclusion**
+- The current `input type` implementation need to be broken up into:
+  - additional `access types` (for `file`, `dir`, ...)
+  - `resource.uploader` transformations (for `helm`)
+
+#### Plugin Type System Reuse
+- The transformation specification implementation is aware of the supported 
+  capabilities. So, depending on the *capability* (such as `type: resource.
+  downloader`), the implementation will use the corresponding *Provider* (e.
+  g. resource repository provider).
+- All types of a particular *capability* share a single common output schema.
+
+**Advantages**
+- We have a single uniform plugin system.
+- *Capabilities* can define a schema of common fields all *types* have to 
+  provide.
+
+**Disadvantages**
+- Additional indirection adds complexity. We have to first select a plugin 
+  registry based on the *capability* and then select the actual plugin based 
+  on the *type*.
+- It is harder to introduce arbitrary plugins as a plugin registry for the
+  corresponding *capability* has to be available.
+
+**Conclusion** 
+- Generally, this additional layer of indirection adds 
+flexibility at the cost of complexity. This is especially visible in simple use 
+cases (capability with a single type). 
+- However, since we already have this 
+plugin system in place, the complexity of maintaining two different plugin 
+systems is even higher.
+
+#### Resource as Atomic Unit in OCM
+- The atomic unit in ocm is **resource** NOT A PLAIN BLOB or ACCESS, kind of 
+  like the atomic unit in kubernetes are pods not containers.
+- While this takes the operations to a higher abstraction level, offering each
+  operation the ability to add or modify the metadata of a resource.
+
+> [!NOTE]
+> In ocm v1, several interfaces were built against plain blobs or access as 
+> the atomic unit. This led to issues, as 
+> - several operations had to be performed twice (digest calculation of a blob) 
+>   or as the metadata could not be passed along.
+> - several extension points were not flexible enough (resource uploader could
+>   not provide the digest of the uploaded blob, as it only returned a blob)
+
+#### Static Typing
+- The example above shows the *outputs* of each transformation as comments. This
+  is purely for illustration purposes. In the actual implementation, every
+  capability has to define a static output schema (kind of like kubernetes 
```

Comment:
I would love to see some more concrete pseudo code to expose the type checking / schema behavior.